flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateLocalitySlotAssigner.java

```diff
@@ -111,8 +117,17 @@ public Collection<SlotAssignment> assignSlots(
 
         final Map<String, ExecutionSlotSharingGroup> groupsById =
                 allGroups.stream().collect(toMap(ExecutionSlotSharingGroup::getId, identity()));
+
+        final Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsPerTaskExecutor =
+                getSlotsPerTaskExecutor(freeSlots);
+        final Collection<? extends SlotInfo> pickedSlots =
+                pickSlotsInMinimalTaskExecutors(
+                        slotsPerTaskExecutor,
+                        allGroups.size(),
+                        getSortedTaskExecutors(freeSlots, slotsPerTaskExecutor, scores));
```

Comment:
`Collection<? extends SlotInfo> pickedSlots = freeSlots`;

`getSlotsPerTaskExecutor and pickSlotsInMinimalTaskExecutors` are only needed when `freeSlots.size > allGroups.size()`, right?

If yes, this optimization can be applied for DefaultAssigner.